fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerScanner.java

```diff
@@ -0,0 +1,346 @@
+/*
+ * OnlineIndexerScanner.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2020 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb;
+
+import com.apple.foundationdb.FDBException;
+import com.apple.foundationdb.Transaction;
+import com.apple.foundationdb.async.AsyncUtil;
+import com.apple.foundationdb.async.MoreAsyncUtil;
+import com.apple.foundationdb.async.RangeSet;
+import com.apple.foundationdb.record.IndexState;
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.RecordCursor;
+import com.apple.foundationdb.record.RecordCursorResult;
+import com.apple.foundationdb.record.logging.KeyValueLogMessage;
+import com.apple.foundationdb.record.logging.LogMessageKeys;
+import com.apple.foundationdb.record.metadata.Index;
+import com.apple.foundationdb.record.provider.foundationdb.synchronizedsession.SynchronizedSessionRunner;
+import com.apple.foundationdb.record.query.plan.synthetic.SyntheticRecordFromStoredRecordPlan;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.protobuf.Message;
+import org.apache.commons.lang3.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * A base class for different types of online indexing scanners.
+ */
+public abstract class OnlineIndexerScanner {
```

Comment:
This still feels like not quite the right abstraction. It's a step in the right direction, for sure, but it still seems off.
Full disclosure, I'm surely affected by my preference against inheritance, and my ideas before I saw your approach.
And perhaps I may be having a desire to over-architect the design based on the models I have in my head.

I think, at least one place this manifests itself is the fact that `OnlineIndexerByRecords` is the only one that can build an individual range, which seems weird, or more, it seems weird that adding it to `OnlineIndexerByIndex` requires any implementation at all. 

I may be oversimplifying, but it seems like having the two classes that currently extend from this as separate `source` classes that don't really know anything about writing index entries, or directly know about the index build state. 


Generally a trick that I like to figure out how to build abstractions when there aren't a lot of implementations, is to think about how to test things.
Specifically of note, right now if you want to test any of the components of `OnlineIndexer` you have to build an index. I'm not saying we should expose all sorts of private members, but a different api might allow us to test these more in isolation. For example, I suspect we could have a simple clean `IndexerSource` interface, and then we could have a whole slew of tests for scanning records, versus indexes, versus multi-type indexes, etc. And all we would have to assert about is that every record is returned, always. And if the Throttle was completely decoupled (as you implied in the javadoc), we could more easily test that with random lambdas that do things like throw various retryable exceptions, etc. And it seems like you should be able to isolate the index writing part, although non-idempotent indexes might make that harder. 


All that being said, as far as this PR goes, I would say just mark everything as INTERNAL, since nobody should really be using these classes directly anyways, and it can be hashed out / cleaned up as more behavior is added / improved.

Reply:
A range, if I understand it correctly, byRecords useless db-ranges that can be considered a set of "(interval]" (opened at one endpoint, closed at the other). When not scanning by db records but in arbitrary order (like a source index), is there any meaningful db-range? 
We can surely use  source index continuation pointers that can be considered as splitting the index into index-ranges, yet any existing 'single range' test would crash if attempting a db-ranges enforcement on a source index scan. What am I missing? 

I agree, of course, with the rest of the comment and will try to improve the code.

Reply:
I suppose we'd have to define the meaning, but I think it would be sensible to say that the definition of the "range" involved depends on the `IndexerSource` in question. So for the by records source, that means primary key range; for the by index source, that means ranges of index entries (or index entry keys).

> We can surely use source index continuation pointers that can be considered as splitting the index into index-ranges,

Kind of. Note that continuations aren't quite the right solution here as splitting the index into ranges accomplishes two things:

1. Being able to resume an index build if one stops in the middle. Continuations would be fine to use here.
2. Being able to split up an index build and farm each piece to its own worker. The logic for that currently only exists for splitting by primary key, but splitting the index up by entry key should be able to accomplish the same thing (I think).

> yet any existing 'single range' test would crash if attempting a db-ranges enforcement on a source index scan

Well, I'm not sure it would crash--it may just have incorrect behavior (including but not limited to not indexing all records). But yeah, it would be bad. That's part of the reason I had initially mentioned in a design document/review comment (I think) that this would require storing in the database what index building strategy was used and then checking it to make sure an index build wasn't incorrectly resumed.

Note that as that is a change that can change what gets stored, it's kind of important we settle on what we want now.

Reply:
One other thing that's unclear to me from this class structure: what's the difference between what goes in `OnlineIndexerCommon` and what goes in the `OnlineIndexerScanner` base class?

Reply:
The common class just holds the setup/metadata (+ few getters) and used by all classes (including the interface). The scanner class and its derivatives are implementing the scan & index algorithm

Reply:
I think @alecgrieser correctly rephrased my thoughts (with valuable elaboration).

In addition to storing the strategy, you would have to store the source index subspace key, and maybe should include the metadata version.

Reply:
Yeah, I think it would be index subspace key and the index's last modified meta-data version

Reply:
> The common class just holds the setup/metadata (+ few getters) and used by all classes (including the interface). The scanner class and its derivatives are implementing the scan & index algorithm

I think it would maybe make sense if this were just a "state" object, then, that packaged up state that needed to be shared, with accessors/modifiers. I think the `getRecordStore()` method being in `Common` is a little strange... I guess it's there because it's used by the throttler? Hmm

Reply:
(did you mean `openRecordStore`?) 
Yes, as a getter, it is a stretch. I can move it to the base class and let the throttle get the `recordStoreBuilder` from common and explicitly open the store (it's a simple one liner).